File: packages/shared/src/types/agent-config.ts
/**
 * Agent Configuration Schema - Shared between Frontend and Backend
 * 
 * Issue #125: Unified schema to prevent frontend-backend contract mismatch
 */

import { z } from 'zod';

/**
 * Agent Configuration Schema
 * 
 * This schema defines the contract for agent runtime configuration
 * that is sent from frontend to backend. Both sides must validate
 * against this schema to ensure type safety.
 */
export const AgentConfigSchema = z.object({
  // Search configuration
  searchMode: z.string().default('hybrid'),
  scoreThreshold: z.number().min(0).max(1).default(0.6),
  topK: z.number().int().positive().default(8),
  docTypes: z.array(z.string()).default([]),
  
  // Agent execution
  maxIterations: z.number().int().positive().max(10).default(3),
  
  // LLM configuration
  llmRoute: z.enum(['auto', 'local', 'deepseek']).default('deepseek'),
  llmProvider: z.string().optional(),  // Issue #125 fix: added missing field
  llmModel: z.string().optional(),
  llmEngine: z.string().optional(),
});

export type AgentConfig = z.infer<typeof AgentConfigSchema>;

/**
 * Resolve LLM provider from route if not explicitly provided
 * 
 * This function provides backward compatibility and sensible defaults
 * when frontend doesn't send llmProvider explicitly.
 */
export function resolveLLMProvider(config: AgentConfig): string {
  if (config.llmProvider) {
    return config.llmProvider;
  }
  
  // Fallback: infer provider from route
  const providerMap: Record<AgentConfig['llmRoute'], string> = {
    'deepseek': 'deepseek',
    'local': 'ollama',
    'auto': 'deepseek',
  };
  
  return providerMap[config.llmRoute];
}

/**
 * Default configuration values
 */
export const DEFAULT_AGENT_CONFIG: AgentConfig = {
  searchMode: 'hybrid',
  maxIterations: 3,
  scoreThreshold: 0.6,
  topK: 8,
  docTypes: [],
  llmRoute: 'deepseek',
  llmProvider: undefined,
  llmModel: 'deepseek-chat',
  llmEngine: 'api',
};

File: src/backend/retrieval-service/domain_models/agent_config.py
"""
Agent Configuration Domain Models

Issue #125: Unified schema for frontend-backend contract.
Extracted from app/api.py to enable lightweight import (no deep service dependencies).

Schema mirrors @rag/shared AgentConfigSchema in packages/shared/src/types/agent-config.ts.
"""

from typing import Optional
from pydantic import BaseModel


PROVIDER_MAP: dict[str, str] = {
    "deepseek": "deepseek",
    "local": "ollama",
    "auto": "deepseek",
}

DEFAULT_PROVIDER = "deepseek"


class AgentRequest(BaseModel):
    """Sync agent request (POST /api/v1/agent)"""
    query: str
    session_id: Optional[str] = None
    max_iterations: int = 3
    llm_route: str = "deepseek"
    llm_provider: Optional[str] = None
    llm_model: Optional[str] = None
    llm_engine: Optional[str] = None

    def resolve_llm_provider(self) -> str:
        """Infer provider from route if not explicitly set (Issue #125)."""
        if self.llm_provider:
            return self.llm_provider
        return PROVIDER_MAP.get(self.llm_route, DEFAULT_PROVIDER)


class AgentStreamRequest(BaseModel):
    """Streaming agent request (POST /api/v1/agent/stream)"""
    query: str
    session_id: Optional[str] = None
    max_iterations: int = 3
    score_threshold: float = 0.60
    top_k: int = 8
    search_mode: str = "hybrid"
    doc_types: list = []
    llm_route: str = "deepseek"
    llm_provider: Optional[str] = None
    llm_model: Optional[str] = None
    llm_engine: Optional[str] = None

    def resolve_llm_provider(self) -> str:
        """Infer provider from route if not explicitly set (Issue #125)."""
        if self.llm_provider:
            return self.llm_provider
        return PROVIDER_MAP.get(self.llm_route, DEFAULT_PROVIDER)


def build_llm_config(request: AgentStreamRequest | AgentRequest) -> dict:
    """Build LLM config dict, always resolving provider (never None)."""
    return {
        "route_mode": request.llm_route,
        "provider": request.resolve_llm_provider(),
        "model": request.llm_model,
        "engine": request.llm_engine,
    }
